### DISCUSSION TITLE
How to ensure Secure Communication between peers

### DISCUSSION THREAD

**Discussion by CJ-404** [libp2p/go-libp2p#2776] [Q&A]
I want to know how to use **TLS** or **Noice** to secure communication. In most places documentation says secure communication can be provide by using ether one of these! but I couldn't find how.

In one [topic in libp2p discussion forums](https://discuss.libp2p.io/t/how-to-use-tls-in-libp2p/975), I saw the peers are verified using there Peer IDs so we don't need to explicitly use TLS. ( connection is secured by default). Is that true?

```
nodeHost, err = libp2p.New(
		libp2p.Identity(prvKey),
		libp2p.ListenAddrs([]multiaddr.Multiaddr(config.ListenAddresses)...),
		// support TLS connections
		libp2p.Security(libp2ptls.ID, libp2ptls.New),
		// support Noise connections
		libp2p.Security(noise.ID, noise.New),
		// support any other default transports (TCP)
		libp2p.DefaultTransports,
		// control too many connections
		libp2p.ConnectionManager(connmgr),
	)
```

Will creating the host like this apply encryption decryption as well? I can't understand it clearly.
**How it know the public key?**
**Can I use a generated keypair to encrypt and decrypt?**

Can someone explain how this works please. 

#Noice #TLS #secureCommunication #encryption #decryption #keyPair #libp2p-go

**Comment by MarcoPolo** [comment]
`libp2p.Identity` lets you use a pregenerated private key. If none is provided libp2p will generate one automatically.

By default both TLS and Noise are enabled. If a peer supports both, the client will prefer TLS.
